File: linguagem_c/week_01/credit/credit.c
#include <stdio.h>
#include <string.h>
#include <ctype.h>

char* validate_credit_card(long long card_number);

int main(){
  char input[32];
  long long card_number;
  int valido;

  do {
    valido = 1;

    printf("Number: ");
    scanf("%31s", input);

    for (int i = 0; input[i] != '\0'; i++) {
      if (!isdigit(input[i])) {
        valido = 0;
        break;
      }
    }

    if (valido) {
      sscanf(input, "%lld", &card_number);
    }
  } while (!valido);


  printf("%s\n", validate_credit_card(card_number));

  return 0;
}

char* validate_credit_card(long long card_number){
  int pick_number = 0;
  int sum = 0;

  long long number = card_number;

  while(number > 0){
    int digit = number % 10;

      if (pick_number)
      {
        digit *= 2;
        if (digit > 9){
          digit -= 9;
        }
      }

    sum += digit;
    number /= 10;
    pick_number = !pick_number;
  }

  if(sum % 10 != 0){
    return "INVALID";
  }

  long long starts = card_number;

  while (starts >= 100){
    starts /= 10;
  }

  if (starts == 34 || starts == 37){
    return "AMEX";
  }
  else if (starts == 51 || starts == 52 || starts == 53 || starts == 54 || starts == 55){
    return "MASTERCARD";
  }
  else
  {
    while (starts >= 10){
      starts /= 10;
    }

    if (starts == 4){
      return "VISA";
    }
    else{
      return "INVALID";
    }
  }

}

File: python/week_06/sentimental-credit/credit.py
def validate_credit_card(card_number: str) -> str:
  total_sum = 0
  reverse_digits = card_number[::-1]

  for i, digit in enumerate(reverse_digits):
    n = int(digit)

    if i % 2 == 1:
      n *= 2
      if n > 9:
        n -= 9

    total_sum += n

  if total_sum % 10 != 0:
    return "INVALID"

  length = len(card_number)

  if length == 15 and card_number.startswith(("34", "37")):
    return "AMEX"
  elif length == 16 and card_number.startswith(("51", "52", "53", "54", "55")):
    return "MASTERCARD"
  elif length in (13, 16) and card_number.startswith("4"):
    return "VISA"
  else:
    return "INVALID"


def main():
  while True:
    card_input = input("Number: ")

    if card_input.isdigit():
      break

  print(validate_credit_card(card_input))

if __name__ == "__main__":
  main()